src/dia/Paper.mjs

```diff
@@ -2234,6 +2261,13 @@ export const Paper = View.extend({
         var localPoint = this.snapToGrid(originalEvent.clientX, originalEvent.clientY);
         var delta = Math.max(-1, Math.min(1, originalEvent.wheelDelta));
 
+        //if( ??? ) {
+            evt.preventDefault(); // shouldn't be calling preventDefault() always!
+            this._mw_evt_buffer.push(originalEvent);
+            this._processMouseWheelEvtBuf();
+            return; //
```

Comment:
Yes, I agree. The thing is that I'm not really sure how can we make this work.

Currently the `_processMouseWheelEvtBuf()` is debounced, which means that the user can't actually call `preventDefault()` for each event. That means that the pinch gesture actually triggers the browser's default behavior (which is to zoom the entire webpage; instead of what the user actually wants: zoom the paper).

![image](https://user-images.githubusercontent.com/89727/187513929-cc94f102-29e3-4e79-a338-38cb4a2acb32.png)

Reply:
This will work for now, right? That is fine by me at this stage.
```js
paper.on('cell:mousewheel blank:mousewheel', evt => evt.preventDefault());
paper.on('TBD_system:scale': (_evt, x, y, scale) => {
  const zoom = this.paperScroller.zoom();
  this.paperScroller.zoom(zoom * scale, { min: 0.2, max: 5, ox: x, oy: y, absolute: true });
});
```
We can then decide if
- we introduce an option to prevent default
- should we trigger `TBD_system:scale` event only if the event was previously prevented,
- should debounce it with `leading: true` and then check if `evt.isDefaultPrevented()` and call `evt.preventDefault()` on each `evt` in buffer.
- other option
 
I am not sure yet.

Reply:
Hold on, I think I got it!! Pushing a new set of commits in ~10min

Reply:
@kumilingus New logic. Pinch gesture is debounced and the events are being canceled (`.preventByDefault()`) because it doesn't make any sense not to do so (pinch gesture will always translate to a "scale" event, which the browser will never handle properly).

If there is a mousewheel event that is not a pinch, it might be either a 2-axis scroll (touchpad) or a regular scroll.

If it's a 2-axis scroll, trigger a `TBD_system:scroll` event. Else, just do what the library was already doing up until now.

How does that sound?

Reply:
It sounds very good :) This is what I would do next.

1. call the new events `paper:pinch` and `paper:pan` (later when we support touch screen gestures they might trigger the same events). If the user decides to draw and resize a rectangle on `paper:pinch`, the `zoom` name would not be precise.

3. `paper:pan` should be triggered even for vertical and horizontal scrolling

```js
paper.on("paper:pan", (evt, x, y) => {
    evt.preventDefault();
    paperScroller.el.scrollLeft += x;
    paperScroller.el.scrollTop += y;
});
```

4. `cell:mousewheel` and `blank:mousewheel` should be triggered with the original `delta`. New `deltaY` it's way to larger then the previous `delta` and it breaks this use-case.
```js
paper.on('blank:mousewheel', (evt, x, y, delta) => {
    evt.preventDefault();
    paperScroller.zoom(delta * 0.2, { min: 0.2, max: 5, grid: 0.2, ox: x, oy: y });
});
```

5. I think we should still make the `preventDefault()` optional (or entire `paper:pinch` branch). It stops the browser zoom with pinch gesture, which could be important accessibility feature in existing applications.

6. Add new events to typescript definition.

7. Write docs.

Reply:
I'm really sorry for taking so long to push the next batch of updates, I have been overwhelmed with work this week. I'll try my best to do some progress with weekend!

Reply:
@kumilingus 

I got some spare time today so I could get some work done! I have several questions:

> 2.

Are you sure? If the user subscribes to both event, how would he/she differentiate them?

> 4.

Actually, the value of `deltaY` is the value that should have been used from the beginning, since that is the actual normalized value across all browsers; `delta` is "wrong". But I do understand (and agree) that using `deltaY` would be kind of a breaking change, so I'll revert that and use the `delta` value 😉.

Btw, you missed the `3.`! 😝

> 5.

This one is tricky. Let's split it into several parts:

> It stops the browser zoom with pinch gesture

Not entirely true. It will stop the pinch gesture only if the gesture is made while the cursor is inside the canvas.

> I think we should still make the preventDefault() optional

This is technically impossible. We're debouncing the `_processMouseWheelEvtBuf`, which means that by the time we've triggered the `paper:pinch` event and have given the user the chance to call `preventDefault()` it would be already too late. 

That said, I can offer you a partial "solution". Whenever we get a pinch event, first check if the `paper:pinch` event has any subscribers. If there are none, just skip the entire code path. (check commit `c66f14d70fda402694e45fa2b14a16c8600cbda4`)

Reply:
(2nd part of my reply)

> 6.

I assume this would go somewhere into JJ+'s repo, right? Looking at the distributable zip I have access to in "My downloads" section, I'd guess that maybe inside `rappid.d.ts` (?), but I'm not sure. Can you give me some tips on this one?

> 7.

I can't find the right place to do this. Will it work if I write the docs "here" (as a comment maybe?) and let you add them to the correct place? Or if you could tell me where should I add the docs, then I could add them.

🙏

Reply:
2.
> Are you sure? If the user subscribes to both event, how would he/she differentiate them

What both events?


Typically, the user wants to enable the panning in all directions. We want to make it as easy as possible:
```js
paper.on("paper:pan", (evt, tx, ty) => {
    evt.preventDefault();
    paperScroller.el.scrollLeft += tx;
    paperScroller.el.scrollTop += ty;
});
```
The current implementation is more like a `paper:diagonal-pan` and not `paper:pan`. The `paper:pan` is a possible replacement for `cell:mouswheel` and `blank:mousewheel`.

And if I need to enable panning I have to do this:
```js
// just a diagonal panning
paper.on("paper:pan", (evt, tx, ty) => {
    evt.preventDefault();
    paperScroller.el.scrollLeft += tx;
    paperScroller.el.scrollTop += ty;
});
// If I want to be explicit, I define horizontal and vertical panning like this:
paper.on("blank:mousewheel", (evt) => {
    evt.preventDefault();
    paperScroller.el.scrollLeft += evt.originalEvent.deltaX;
    paperScroller.el.scrollTop += evt.originalEvent.deltaY;
});
```

3. 🤦  😄

4. 👍 

5.
    The canvas usually covers most of the screen. That's was my reasoning. It would be technically possible to prevent default with a paper option. `if (paper.options.preventDefaultPinch)  evt.preventDefault();`.
    But your solution with listener checks is much better. I wouldn't call it a partial solution. We're just making life easier for developers by doing it automatically. I really like it.
    
    I can imagine we could also add `hasListener(eventName): boolean` method to `mvc.View`. Or later to `mvc.Events` mixin once it exists (https://backbonejs.org/#Events).

6. No, it goes [here](https://github.com/clientIO/joint/blob/master/types/joint.d.ts#L1315-L1388) right in JJ repo.

7. And the documentation goes [here](https://github.com/clientIO/joint/blob/master/docs/src/joint/api/dia/Paper/events.html). You can build and run the docs with `npx grunt`, `http-server`, `open http://localhost:8080/build/docs/joint.html`.

Reply:
> 2

Oh, I think I gotcha! Check commit `86ce1c2a6e1aeea7c3c0352b2d70c45db0645781 `

> 5

Oh, phew! I'm glad you like that solution. I though you'd reject it because it's kind of hacky (I'm accessing a private variable of backbone).

> 6 / 7

✅